### PR TITLE
Adding Drop Index

### PR DESCRIPTION
**Pull request by pholanda**

### CODE REVIEW

File: src/parser/postgresparser.cpp [hunk 1]
```diff
@@ -1122,6 +1124,15 @@ parser::DropStatement *PostgresParser::DropTriggerTransform(DropStmt *root) {
   return result;
 }
 
+parser::DropStatement *PostgresParser::DropIndexTransform(DropStmt *root) {
+  auto res = new DropStatement(DropStatement::EntityType::kIndex);
```
Comment (haojin2) [inline src/parser/postgresparser.cpp]: Please use "result" as the name of the variable that you will be returning for consistency with all other functions within this file.

File: src/executor/drop_executor.cpp [hunk 2]
```diff
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include <include/catalog/index_catalog.h>
```
Comment (apavlo) [inline src/executor/drop_executor.cpp]: Use formatter.

File: src/executor/drop_executor.cpp [hunk 3]
```diff
@@ -94,6 +95,22 @@ bool DropExecutor::DExecute() {
       }
       break;
     }
+    case DropType::INDEX: {
+      std::string index_name = node.GetIndexName();
+      auto current_txn = context_->GetTransaction();
+      auto index =  catalog::IndexCatalog::GetInstance()->GetIndexObject(index_name,current_txn);
+      bool result = catalog::IndexCatalog::GetInstance()->DeleteIndex(index->GetIndexOid(),current_txn);
+      if (result){
```
Comment (apavlo) [inline src/executor/drop_executor.cpp]: Use formatter.

File: src/include/parser/postgresparser.h [hunk 4]
```diff
@@ -220,6 +220,9 @@ class PostgresParser {
   // transform helper for drop trigger statement
   static parser::DropStatement *DropTriggerTransform(DropStmt *root);
 
+  // tranform helper for drop index statement
+  static parser::DropStatement *DropIndexTransform(DropStmt *root);
```
Comment (apavlo) [inline src/include/parser/postgresparser.h]: See #1011

File: src/executor/drop_executor.cpp [hunk 5]
```diff
@@ -94,6 +95,22 @@ bool DropExecutor::DExecute() {
       }
       break;
     }
+    case DropType::INDEX: {
```
Comment (apavlo) [inline src/executor/drop_executor.cpp]: Is this transactional? How are you invalidating the cached query plans?

File: src/include/planner/drop_plan.h [hunk 6]
```diff
@@ -50,9 +50,11 @@ class DropPlan : public AbstractPlan {
 
   std::string GetTableName() const { return table_name; }
 
-  std::string GetTriggerName() const { return trigger_name; }
+    std::string GetTriggerName() const { return trigger_name; }
```
Comment (apavlo) [inline src/include/planner/drop_plan.h]: Run formatter.

**Comment by apavlo** [comment]
Thanks for sending this in. @poojanilangekar will coordinate with @ChTimTsubasa to make sure that this works with our new plan cache.

**Comment by pholanda** [comment]
@apavlo Left the code working after the  PR-1011 and executed the clang-formatter. Let me know If I can be of any more help.

**Comment by apavlo** [comment]
@ChTimTsubasa I was thinking that we could merge this and then deal with invalidating the plan cache separately. What do you think?

File: src/executor/drop_executor.cpp (hunk 2 above)
Comment (apavlo) [reply src/executor/drop_executor.cpp]: Actually, you have to manually fix this. It should be:

`#include "catalog/index_catalog.h"`

And it should come after "executor/drop_executor.h".

File: src/executor/drop_executor.cpp [hunk 7]
```diff
@@ -10,9 +10,9 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include <include/catalog/index_catalog.h>
 #include "executor/drop_executor.h"
 
+#include "include/catalog/index_catalog.h"
```
Comment (apavlo) [inline src/executor/drop_executor.cpp]: `"include/catalog/index_catalog.h"` -> `"catalog/index_catalog.h"`

Sorry for being pedantic. The new formatter from @tcm-marcel will do this automatically (#991)

**Comment by ChTimTsubasa** [comment]
@apavlo I agree with you. We can merge this one and the drop first, and then I would insert those lines for plan-cache notification. It's relatively easy.

**Comment by apavlo** [comment]
@pholanda Actually, we can't merge this without test cases. Create a new test case called `DropIndexTest` here:

https://github.com/cmu-db/peloton/blob/master/test/sql/drop_sql_test.cpp

It can work just like `DropTableTest`:

1. Create the table 
2. Create the index
3. Drop the index
4. Check the catalog to make sure that the index no longer exists

File: src/planner/drop_plan.cpp [hunk 8]
```diff
@@ -40,16 +40,22 @@ DropPlan::DropPlan(parser::DropStatement *parse_tree) {
...
       table_name = std::string(parse_tree->table_name_of_trigger);
       trigger_name = std::string(parse_tree->trigger_name);
       drop_type = DropType::TRIGGER;
       break;
     }
-    default: {
-      LOG_ERROR("Not supported Drop type");
+    case parser::DropStatement::EntityType::kIndex: {
+      database_name = parse_tree->GetDatabaseName();
+      table_name = std::string(parse_tree->table_name_of_trigger);
```
Comment (ChTimTsubasa) [inline src/planner/drop_plan.cpp]: You are not using table_name or database_name in the executor. Delete them.

Actually, I am not sure if database_name is useful. In Postgres, index name is not visible across databases, but our catalog does not expose the database name in the API. @mengranwo Can you explain how the index is implemented in Peloton catalog?

File: src/executor/drop_executor.cpp [hunk 9]
```diff
@@ -139,5 +144,21 @@ bool DropExecutor::DropTrigger(const planner::DropPlan &node,
...
+bool DropExecutor::DropIndex(const planner::DropPlan &node,
+                             concurrency::TransactionContext *txn) {
+  std::string index_name = node.GetIndexName();
+  auto index_object =
+      catalog::IndexCatalog::GetInstance()->GetIndexObject(index_name, txn);
+  ResultType result = catalog::Catalog::GetInstance()->DropIndex(
+      index_object->GetIndexOid(), txn);
+  txn->SetResult(result);
+  if (txn->GetResult() == ResultType::SUCCESS) {
+    LOG_TRACE("Dropping Index Succeeded!");
```
Comment (poojanilangekar) [inline src/executor/drop_executor.cpp]: Please add the index_oid in the debug statement.

File: src/include/parser/postgresparser.h [hunk 10]
```diff
@@ -180,19 +180,20 @@ class PostgresParser {
 
   /**
    * @brief transform helper for create database statement
-   * 
+   *
    * @param Postgres CreateDatabaseStmt parsenode
    * @return a peloton CreateStatement node
    */
-  static parser::SQLStatement *CreateDatabaseTransform(CreateDatabaseStmt *root);
+  static parser::SQLStatement *CreateDatabaseTransform(
```
Comment (poojanilangekar) [inline src/include/parser/postgresparser.h]: Please remove this formatting change. Also remove similar changes elsewhere.

File: src/parser/postgresparser.cpp [hunk 11]
```diff
@@ -294,10 +294,12 @@ expression::AbstractExpression *PostgresParser::ColumnRefTransform(
                 ->val.str));
       } else {
         result = new expression::TupleValueExpression(
-            std::string((reinterpret_cast<value *>(
-                             fields->head->next->data.ptr_value))->val.str),
-            std::string((reinterpret_cast<value *>(
-                             fields->head->data.ptr_value))->val.str));
+            std::string(
```
Comment (poojanilangekar) [inline src/parser/postgresparser.cpp]: Please revert the formatting change here and in other parts of the file.

**Comment by apavlo** [comment]
I will merge this next when the build passes.

**Comment by apavlo** [comment]
Can you resolve the conflicts and then I will merge. Sorry about that

**Comment by apavlo** [comment]
@pholanda Sorry I should have been more clear. You need to rebase too.

**Comment by apavlo** [comment]
I rebased this correctly and merged manually.

**Pull request closed without merging**